read_header.hpp

```diff
@@ -435,3 +488,115 @@ void validate(const std::vector<std::vector<std::vector<std::vector<double>>>>&
     }
 }
 
+
+//Calculating the temperature:
+#include <boost/math/constants/constants.hpp>
+#include <boost/math/tools/roots.hpp>
+//#include <boost/math/interpolators/bilinear.hpp>
+double bilinear_interpolate(double x, double y,
+                            const std::vector<double>& x_vals,
+                            const std::vector<double>& y_vals,
+                            const std::vector<std::vector<double>>& f_vals) {
+    size_t i = std::upper_bound(x_vals.begin(), x_vals.end(), x) - x_vals.begin() - 1;
+    size_t j = std::upper_bound(y_vals.begin(), y_vals.end(), y) - y_vals.begin() - 1;
+
+    double x1 = x_vals[i], x2 = x_vals[i+1];
+    double y1 = y_vals[j], y2 = y_vals[j+1];
+
+    double Q11 = f_vals[i][j];
+    double Q21 = f_vals[i+1][j];
+    double Q12 = f_vals[i][j+1];
+    double Q22 = f_vals[i+1][j+1];
+
+    double denom = (x2 - x1) * (y2 - y1);
+    return (Q11 * (x2 - x) * (y2 - y) +
+            Q21 * (x - x1) * (y2 - y) +
+            Q12 * (x2 - x) * (y - y1) +
+            Q22 * (x - x1) * (y - y1)) / denom;
+}
+
+std::function<double(double, double)>
+make_mu_interpolator(const std::vector<double>& Log_n_H_vals,
+                    const std::vector<double>& T_vals,
+                    const std::vector<std::vector<double>>& mu_table) {
+    return [=](double n_H, double T) {
+        return bilinear_interpolate(std::log10(n_H), T, Log_n_H_vals, T_vals, mu_table);
+    };
+}
+
+std::vector<double> read_vector_csv(const std::string& filename) {
+    std::vector<double> vec;
+    std::ifstream file(filename);
+    std::string line;
+    while (std::getline(file, line)) {
+        std::stringstream ss(line);
+        double val;
+        while (ss >> val) {
+            vec.push_back(val);
+            if (ss.peek() == ',') ss.ignore();
+        }
+    }
+    return vec;
+}
+
+std::vector<std::vector<double>> read_matrix_csv(const std::string& filename) {
+    std::vector<std::vector<double>> matrix;
+    std::ifstream file(filename);
+    std::string line;
+    
+    while (std::getline(file, line)) {
+        std::vector<double> row;
+        std::stringstream ss(line);
+        std::string val;
+        
+        while (std::getline(ss, val, ',')) {
+            row.push_back(std::stod(val));
+        }
+        matrix.push_back(row);
+    }
+
+    return matrix;
+}
+
+
+
+// Constants
+constexpr double k_B = 1.380649e-16; // Boltzmann constant in erg/K
+constexpr double m_p = 1.6726219e-24; // Proton mass in g
+constexpr double X = 0.76; // Hydrogen mass fraction
+constexpr double z_val = 0.0; // Redshift value, can be adjusted based on the simulation
+
+// Solving the equation (gamma-1)e_int= n * k_B * T / (mu * m_p) for T
+//i.e., (gamma-1) * e_int*mu(n_H,T)*m_p /(rho*k_B) -T=0
+
+
+
+double residual_function(double T, double e_int, double rho, std::vector<double> Temperatures, 
```

Comment:
If this is just for validation and if you've convinced yourself that the temperature being calculated is correct, you can remove this.